src/encoder.rs

```diff
@@ -807,140 +808,580 @@ impl<W: Write> Drop for Writer<W> {
+    ///
+    /// *Note that if the first frame uses [`DisposeOp::Previous`]
+    /// it will be treated as [`DisposeOp::Background`].*
+    ///
+    /// This method will return an error if the image is not animated.
+    ///
+    /// [`DisposeOp`]: ../common/enum.BlendOp.html
+    /// [`DisposeOp::Previous`]: ../common/enum.BlendOp.html#variant.Previous
+    /// [`DisposeOp::Background`]: ../common/enum.BlendOp.html#variant.Background
+    pub fn set_dispose_op(&mut self, op: DisposeOp) -> Result<()> {
+        if let Some(ref mut fctl) = self.fctl {
+            fctl.dispose_op = op;
+            Ok(())
+        } else {
+            Err(EncodingError::Format(FormatErrorKind::NotAnimated.into()))
         }
     }
 
     pub fn finish(mut self) -> Result<()> {
+        if !self.end {
+            let err = FormatErrorKind::MissingFrames.into();
+            return Err(EncodingError::Format(err));
+        } else if self.to_write > 0 {
+            let err = FormatErrorKind::MissingData(self.to_write).into();
+            return Err(EncodingError::Format(err));
+        }
         // TODO: call `writer.finish` somehow?
         self.flush()?;
         Ok(())
     }
+
+    /// Flushes the buffered chunk, checks if it was the last frame,
+    /// writes the next frame header and gets the next frame scanline size
+    /// and image size.
+    fn new_frame(&mut self) -> Result<()> {
+        let mut wrt = match self.writer.take() {
+            Wrapper::Chunk(wrt) => wrt,
+            _ => unreachable!(),
+        };
+        wrt.flush()?;
+        if let Some(fctl) = self.fctl {
+            wrt.set_fctl(fctl);
+        }
+        let (scansize, size) = wrt.next_frame_info();
+        self.line_len = scansize;
+        self.to_write = size;
+        wrt.writer.written += 1;
+        wrt.write_header()?;
+        self.end = wrt.writer.written + 1 == wrt.writer.max_frames();
+        self.writer = Wrapper::Zlib(ZlibEncoder::new(wrt, self.compression.to_options()));
+        Ok(())
+    }
 }
 
-impl<W: Write> Write for StreamWriter<'_, W> {
-    fn write(&mut self, mut buf: &[u8]) -> io::Result<usize> {
-        let written = buf.read(&mut self.curr_buf[self.index..])?;
+impl<'a, W: Write> Write for StreamWriter<'a, W> {
+    fn write(&mut self, mut data: &[u8]) -> io::Result<usize> {
+        if data.is_empty() {
+            return Ok(0);
+        }
+
+        match self.writer {
+            Wrapper::Chunk(_) => self.new_frame()?,
+            Wrapper::Zlib(_) => {}
+            Wrapper::None => unreachable!(),
+        }
+
+        if self.to_write == 0 {
+            if self.end {
+                let err = FormatErrorKind::EndReached.into();
+                return Err(EncodingError::Format(err).into());
+            }
+            self.writer = match self.writer.take() {
+                Wrapper::Zlib(wrt) => Wrapper::Chunk(wrt.finish()?),
+                _ => unreachable!("invalid writer state"),
+            };
```

Comment:
So if this errors in `wrt.finish()?`, doesn't it leave `self.writer` in the state `Wrapper::None`. Then I call `write` again and we enter the `unreachable` above.

Reply:
Yes you are totally right, I overlooked this line, but it seems to me that if this happens there is nothing to do, we wouldn't be able to continue because the `ChunkWriter` would get lost. Maybe I should add a variant to the `Wrapper` enum like `Unrecoverable` that would return an `EncodingError` every time `write` is called after that error. Maybe `None` can be used for that but it might be better to enforce the fact that `None` is used only as a temporary state